### PR TITLE
Allow force enabling jemalloc on macOS

### PR DESCRIPTION
**Pull request by benesch**
Twiddle our crates and Cargo configuration for activating jemalloc to achieve the following ideal developer experience:

  * `--no-default-features` unequivocally disables jemalloc, regardless of platform.
  * `--features=jemalloc` unequivocally enables jemalloc, regardless of platform.
  * `--default-features` chooses the best allocator for the platform: the system allocator on macOS and jemalloc on Linux.

Since Cargo doesn't explicitly support target-specific features, the trick is to introduce a crate of indirection, as described in [[0]].

Here's how it works:

  * The `mz-prof`, `mz-prof-http`, and `mz-alloc` crates have a straightforward setup: they expose a `jemalloc` feature and conditionally enable or disable jemalloc features based on whether the `jemalloc` feature is provided or not.

  * The `mz-alloc-default` crate depends on the `mz-alloc` crate and enables the best features for the platform: `jemalloc` on Linux, and no additional features on macOS. You can think of this crate as unconditionally enabling the best allocator for the platform.

  * The `mz-environmentd` and `mz-clusterd` crates unlock the conditional behavior. Both crates:
      1. depend on `mz-alloc-default` by default, and
      2. expose a `jemalloc` feature that force enables the `mz-alloc/jemalloc` feature.

The motivation for all of this is to allow macOS developers to enable jemalloc upon request, which is useful for certain kinds of testing.

Alternative to #27041.
Closes #27041.

[0]: https://github.com/rust-lang/cargo/issues/1197#issuecomment-1837113787

### Motivation

  * This PR adds a feature requested on Slack.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

File: ci/test/lint-deps.sh [hunk 1]
```diff
@@ -53,7 +61,9 @@ function deps() {
     # output if there is no dependency to any of the specified packages.
     for crate in "${crates[@]}"; do
         # Gather data for the current target, reverse lines, find the dependency hierarchy to the root, reverse lines.
-        output=$(cargo tree --workspace --format ':{lib}:{f}' \
+        output=$(cargo tree \
+            $(printf -- "-p %s " "${entrypoints[@]}") \
```
Comment (antiguru) [inline ci/test/lint-deps.sh:65]: Oh neat, this didn't work when I wrote the script!

**Review by antiguru — APPROVED** [review]
Looks fine, thank you!

File: src/alloc-default/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+# Consider re-evaluating if jemalloc's macOS support improves.
+#
+# [0]: https://github.com/jemalloc/jemalloc/issues/26
+# [1]: https://github.com/jemalloc/jemalloc/issues/843
+# [2]: https://github.com/jemalloc/jemalloc/issues/1467
+#
+# Furthermore, as of August 2022, some engineers are using profiling tools, e.g.
+# `heaptrack`, that only work with the system allocator on macOS.
+[target.'cfg(not(target_os = "macos"))'.dependencies]
+mz-alloc = { path = "../alloc", features = ["jemalloc"] }
```
Comment (guswynn) [inline src/alloc-default/Cargo.toml:31]: fascinating that cargo correctly overrides the default dep here!

**Review by guswynn — APPROVED** [review]
> Since Cargo doesn't explicitly support target-specific features, the trick is to introduce a crate of indirection, as described in [https://github.com/rust-lang/cargo/issues/1197#issuecomment-1837113787].

i thought about this but couldnt figure out how it'd work! neat!! 

This is awesome, thanks!!

File: src/alloc-default/Cargo.toml (hunk 2 above)
Comment (benesch) [reply src/alloc-default/Cargo.toml:31]: Yeah, when they implemented [target-specific dependencies](https://doc.rust-lang.org/cargo/reference/resolver.html#feature-resolver-version-2), they actually did all the hard work to support target-specific features, too. This is why they rolled out that `resolver = "2"` thing, if you remember. The second version of the resolver is the good one that gets this right.

What they didn't do is the last mile to allow saying `[target.'cfg(...)'.features]`. Which is a little tragic. All that hard work just to miss the last little bit of plumbing.

The good news is that I'm convinced every target-specific feature _can_ be represented as a target-specific dependency with one or more indirection crates. Just can be annoyingly tricky to figure out how to do the transformation.

**Pull request merged**